torch/distributed/_tensor/random.py

```diff
@@ -301,3 +261,158 @@ def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
+        elif isinstance(_parallel_rng_style, DTensorRNG):
+            _parallel_rng_style._manual_seed(seed)
+        else:
+            raise RuntimeError("Unknown parallel RNG style!")
+
+
+# TODO: change this to ABC
+class ParallelRNGStyle:
+    def __init__(self):
+        pass
+
+    # override this function to define custom parallel random number generation
+    # to execute DTensor computation
+    @contextlib.contextmanager
+    def _parallel_region(self):
+        pass
+
+class DTensorRNG(ParallelRNGStyle):
+    # The default parallel random number generation style
+    def __init__(self):
+        super().__init__()
+
+    def _manual_seed(self, parallel_seed: int) -> None:
+        # do not pollute torch's default RNG state
+        device = torch.cuda.current_device()
+        with torch.random.fork_rng([device]):
+            torch.cuda.manual_seed(parallel_seed)
+            _rng_tracker.set_rng_state("parallel-rng", torch.cuda.get_rng_state())
+
+    @contextlib.contextmanager
+    def _parallel_region(self, spec: DTensorSpec):
+        # check if the parallel rng state has been synchronized or not
+        if not self.rng_state_is_sync("parallel-rng"):
+            raise RuntimeError(
+                "DTensorRNG requires random seed to be synchronized"
+                "before entering into a parallel region!"
+            )
+
+        # TODO: rewrite this set/restore logic using fork_rng()?
+        # fork rng state first since parallel RNG state should be used instead
+        old_state = torch.cuda.get_rng_state()  # a 16-byte ByteTensor
+        torch.cuda.set_rng_state(self.rng_states["parallel-rng"])
+        # set offset before entering the parallel region
+        mesh = spec.mesh
+        old_offset = _get_rng_offset(mesh)
+        set_pre_op_offset(spec)
+        try:
+            yield
+        finally:
+            # update offset to synchronize among ranks
+            set_post_op_offset(spec, old_offset)
+            _rng_tracker.set_rng_state("parallel-rng", torch.cuda.get_rng_state())
+            torch.cuda.set_rng_state(old_state)
+
+
+class MegatronRNG(ParallelRNGStyle):
+    def __init__(self):
+        super().__init__()
+
+    def _manual_seed(
+        self,
+        device_mesh: DeviceMesh,
+        default_seed: int = 1234,
+        tp_dim: int = 0,
+    ):
+        # Megatron maintains two sets of RNG states:
+        #   1. default state: torch's cuda RNG state. This is set to the same
+        #   across all ranks unless `data_parallel_random_init` is explicitly
+        #   set to True. Megatron uses this RNG state for non-tensor-parallel
+        #   regions.
+        #   2. tensor model parallel state: separate from the default state
+        #   and is only used in tensor-parallel regions. This is set to the same
+        #   across tensor parallel groups (i.e. along non-tp dimension) but different
+        #   within a tensor parallel group.
```

Comment:
What is the UX for 2D parallel? Can you provide a user flow here? Because in 2D parallel, you need to differentiate whether it's TP or SP and whether it's within TP or not, right? How are going to handle that?

Reply:
basically to support SP we only need to add a `DistributedRNGStyle` that implements the corresponding `distribute_region` context and call it within SP's API:

```
def SomeSPApi():
    with SequenceParallelRNGStyle.distribute_region():
        # perform SP computation here
```

Reply:
Well, if for initialization, user can do this one-time thing, but we don't want to call cx within forward code right?